### PR TITLE
 Datastream Id Override and Datastream Config Override Implementation

### PR DESCRIPTION
**Pull request by addb**
## Description
Similar to iOS PR https://github.com/adobe/aepsdk-edge-ios/pull/398

Javadoc folder looks like this:

Screenshot 2023-10-13 at 9 05 08 AM

Screenshot 2023-10-13 at 9 02 53 AM

----
#### Test Scope
| # |Test|
| :-- | :-- |
| 1 | `Unit tests` - modular tests per method+all others mocked, exercising the code with good code coverage |
| 2 | `Functional tests` - e2e tests with public APIs+mocked network, verify the network requests |
|3 | `Integration tests` - e2e tests with public APIs+network calls, verify the network responses |

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/edge/build.gradle [hunk 1]
```diff
@@ -74,6 +81,11 @@ spotless {
         endWithNewline()
         licenseHeaderFile "../../config/formatter/adobe.header.txt"
     }
+    kotlin {
+        target "src/*/java/**/*.kt"
+        ktlint('0.41.0')
```
Comment (emdobrin) [inline code/edge/build.gradle:86]: - core and other extensions use 0.47.0 for ktlint
- also don't forget to include the kotlin classes to the coverage report for both unit and functional tests

File: Makefile [hunk 2]
```diff
@@ -25,7 +25,7 @@ AAR_FILE_DIR =  $(ROOT_DIR)/code/$(PROJECT_NAME)/build/outputs/aar
 
 init:
 	git config core.hooksPath .githooks
-	
+
 clean:
```
Comment (addb) [inline Makefile:29]: @praveek and @yangyansong-adbe Can you guys help with reviewing changes for dokka. I added a first Kotlin file in the source for Edge and have made changes for that in
`Makefile`, `build.gradle` for the project (edge-sdk), `build.gradle` for edge module (edge). Did not update the Github workflow.

Comment (yangyansong-adbe) [reply Makefile:29]: @addb  the dokka setting looks good to me. Before merging the PR, please run the `javadocJar` command locally and check the generated HTML files, in case there are some bad links or missing API documentation.

Comment (addb) [reply Makefile:29]: Thanks @yangyansong-adbe

File: code/edge/src/main/java/com/adobe/marketing/mobile/SDKConfig.kt [hunk 3]
```diff
@@ -11,15 +11,15 @@
 
 package com.adobe.marketing.mobile
 
-internal data class SDKConfig(val datastream: Datastream) {
+data class SDKConfig(val datastream: Datastream) {
```
Comment (praveek) [inline code/edge/src/main/java/com/adobe/marketing/mobile/SDKConfig.kt:14]: Can you move these new public classes inside edge subpackage?

File: code/edge/src/test/java/com/adobe/marketing/mobile/RequestBuilderTest.java [hunk 4]
```diff
@@ -315,6 +317,74 @@ public void getPayloadWithExperienceEvents_addsKonductorConfigWithStreaming_when
 		assertEquals("\n", payloadMap.get("meta.konductorConfig.streaming.lineFeed"));
 	}
 
+	@Test
+	public void getPayloadWithExperienceEvents_setsSdkConfigMeta_whenSdkConfigPresent() throws Exception {
+		List<Event> events = getSingleEvent(getExperienceEventData("value", "5dd603781b95cc18a83d42ce"));
```
Comment (emdobrin) [inline code/edge/src/test/java/com/adobe/marketing/mobile/RequestBuilderTest.java:322]: nit: can we skip the datasetId setting for these tests unless the intention is to test both config overrides + dataset id

File: code/edge/src/test/java/com/adobe/marketing/mobile/RequestBuilderTest.java [hunk 5]
```diff
@@ -315,6 +317,74 @@ public void getPayloadWithExperienceEvents_addsKonductorConfigWithStreaming_when
...
+		requestBuilder.addSdkConfig(new SDKConfig(new Datastream("OriginalDatastreamId")));
+		JSONObject payload = requestBuilder.getPayloadWithExperienceEvents(events);
+
+		assertNotNull(payload);
+		assertNumberOfEvents(payload, 1);
+
+		Map<String, String> payloadMap = new HashMap<>();
+		addKeys("", new ObjectMapper().readTree(payload.toString()), payloadMap);
```
Comment (emdobrin) [inline code/edge/src/test/java/com/adobe/marketing/mobile/RequestBuilderTest.java:330]: cc @timkimadobe looks like this addKeys helper for unit tests is similar to the flattening one

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java [hunk 6]
```diff
@@ -1287,6 +1216,260 @@ public void setLocationHint(final String hint, final int ttlSeconds) {
...
+		JSONObject metaJson = requestJson.getJSONObject("meta");
+		assertNotNull(metaJson);
+		JSONObject configOverridesJson = metaJson.getJSONObject("configOverrides");
+		assertNotNull(configOverridesJson);
+
+		assertEquals("value", configOverridesJson.get("key"));
+	}
+
+	@Test
+	public void testProcessHit_experienceEvent_whenDatastreamIdAndConfigOverridesSet_sendsSdkConfigAndConfigOverrides()
```
Comment (emdobrin) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1396]: @addb I see an overlap between unit and functional tests especially on the request payload, can you summarize what is being tested and what level?

File: code/edge/build.gradle [hunk 7]
```diff
@@ -74,37 +82,40 @@ spotless {
...
 
-tasks.register('javadocPublish', Jar) {
-    from javadocPublic
-    archiveClassifier.set('javadoc')
+dokkaJavadoc.configure {
+    dokkaSourceSets {
+        named("main") {
+            noAndroidSdkLink.set(false)
+            perPackageOption {
+                matchingRegex.set(".*\\.internal.*") // proper setting
```
Comment (emdobrin) [inline code/edge/build.gradle:104]: what does proper setting mean? 😄

File: code/edge/build.gradle [hunk 8]
```diff
@@ -74,37 +82,40 @@ spotless {
...
-    options.memberLevel = JavadocMemberLevel.PRIVATE
+tasks.register('javadocJar', Jar) {
+    dependsOn dokkaJavadoc
+    archiveClassifier.set('javadoc')
+    from dokkaJavadoc.outputDirectory
 }
 
-tasks.register('javadocPublic', Javadoc) {
-    destinationDir = reporting.file("javadocPublic")
-    options.memberLevel = JavadocMemberLevel.PUBLIC
```
Comment (emdobrin) [inline code/edge/build.gradle:102]: how is the javadocMemberLevel especially for the public lib managed now?

File: code/edge/build.gradle (hunk 7 above)
Comment (addb) [reply code/edge/build.gradle:104]: I got that setting from Core/Analytics 😅

File: code/edge/build.gradle (hunk 8 above)
Comment (addb) [reply code/edge/build.gradle:102]: With the current change we are only generating docs for public classes.

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java (hunk 6 above)
Comment (addb) [reply code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1396]: I though of adding these unit tests since hit processor is responsible for adding the overrides payload and also passes datastream Id for the URL. So it felt like adding tests to verify those things here, which I agree would also be captured in the functional tests.

File: code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.adobe.marketing.mobile.edge.Datastream
+import com.adobe.marketing.mobile.edge.SDKConfig
+import org.junit.Assert.assertEquals
+import org.junit.Test
+
+class SDKConfigTest {
+
+    @Test
+    fun test_datastream_toMap_happy() {
```
[the remark this answers is not in the document]
Comment (addb) [reply code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt:22]: Cannot pass null to this since we expect non-optional Datastream object and Original string .

File: code/edge/build.gradle (hunk 7 above)
Comment (addb) [reply code/edge/build.gradle:104]: If we want any public class which is in subpackage internal would not generate docs for that class. I will remove the proper setting comment.

File: code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (emdobrin) [reply code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt:22]: How about empty strings/maps

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java (hunk 6 above)
Comment (emdobrin) [reply code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1396]: Unit tests is a great candidate for the heavy testing for good coverage, and on top of that we should minimize the amount of test duplication at functional/integration testing. 
@addb @kevinlind what test schema (unit > functional > integration) do you recommend for this feature?

Comment (addb) [reply code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1396]: @emdobrin @kevinlind 
Should I add the functional test in this PR, so it helps with the decision about @emdobrin's comment, as to what to test where?

File: code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (addb) [reply code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt:22]: I added empty string for original which result in empty string returned in the map, since we have validation on the class using this the SDKConfig.

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java (hunk 6 above)
Comment (addb) [reply code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1396]: I think we should keep all the unit tests. We can reduce the functional test and integration test to 1 and test everything(datastream id override + datastream config override) in that one test. Unit tests are fast and whereas functional and integration tests are slow and would increase build time as well.

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java [hunk 10]
```diff
@@ -1287,6 +1217,256 @@ public void setLocationHint(final String hint, final int ttlSeconds) {
...
+		configOverrides.put("key", "value");
+
+		DataEntity dataEntity = new EdgeDataEntity(
+			getExperienceEventWithConfig("testDatastreamId", null),
+			edgeConfig,
+			identityMap
+		)
+			.toDataEntity();
+		// test
+		hitProcessor =
```
Comment (kevinlind) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1283]: Why do you need to recreate the hit processor here? Can you add a note stating the differences with the default?

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java [hunk 11]
```diff
@@ -1287,6 +1217,256 @@ public void setLocationHint(final String hint, final int ttlSeconds) {
...
+						// not called by hit processor
+					}
+				}
+			);
+		assertProcessHitResult(dataEntity, true);
+
+		ArgumentCaptor<String> payloadCaptor = ArgumentCaptor.forClass(String.class);
+		verify(mockEdgeNetworkService, times(1))
+			.doRequest(
+				anyString(),
```
Comment (kevinlind) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1376]: Can you also validate that the config ID in the URL is as expected and not overridden?

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java [hunk 12]
```diff
@@ -1287,6 +1217,256 @@ public void setLocationHint(final String hint, final int ttlSeconds) {
...
+
+	@Test
+	public void testProcessHit_experienceEvent_whenDatastreamIdOverrideSet_sendsSdkConfig() throws Exception {
+		mockNetworkServiceResponse("https://test.com", new RetryResult(EdgeNetworkService.Retry.NO));
+
+		Map<String, Object> configOverrides = new HashMap<>();
+		configOverrides.put("key", "value");
+
+		DataEntity dataEntity = new EdgeDataEntity(
+			getExperienceEventWithConfig("testDatastreamId", null),
```
Comment (kevinlind) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1277]: Repeat this test with an empty string for the datastream ID and validate the original datastream id is used and no SDKConfig.datastream is set in the request.

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java [hunk 13]
```diff
@@ -1287,6 +1217,256 @@ public void setLocationHint(final String hint, final int ttlSeconds) {
...
+		JSONObject metaJson = requestJson.getJSONObject("meta");
+		JSONAsserts.assertExactMatch(expectedSdkConfigJSON, metaJson);
+	}
+
+	@Test
+	public void testProcessHit_experienceEvent_whenDatastreamConfigOverridesSet_sendsConfigOverrides()
+		throws Exception {
+		mockNetworkServiceResponse("https://test.com", new RetryResult(EdgeNetworkService.Retry.NO));
+
+		Map<String, Object> configOverrides = new HashMap<>();
```
Comment (kevinlind) [inline code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java:1338]: Repeat this test with an empty configOverrides map and verify the request payload does not contain the overridden configs.

File: code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (emdobrin) [reply code/edge/src/test/java/com/adobe/marketing/mobile/SDKConfigTest.kt:44]: this goes deep into the business logic, my comment was for the unit tests coverage in this particular class

**Review by kevinlind — APPROVED** [review]

**Review by emdobrin — APPROVED** [review]

File: code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java (hunk 6 above)
Comment (emdobrin) [reply code/edge/src/test/java/com/adobe/marketing/mobile/EdgeHitProcessorTests.java]: Thanks for all the updates on the tests @addb , looks 💯 !
